extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/GradleBuildServerBuildSupport.java

```diff
@@ -0,0 +1,238 @@
+import org.eclipse.jdt.core.IClasspathAttribute;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.ProjectUtils;
+import org.eclipse.jdt.ls.core.internal.ResourceUtils;
+import org.eclipse.jdt.ls.core.internal.managers.IBuildSupport;
+
+import ch.epfl.scala.bsp4j.BuildServer;
+import ch.epfl.scala.bsp4j.BuildTarget;
+import ch.epfl.scala.bsp4j.BuildTargetTag;
+import ch.epfl.scala.bsp4j.OutputPathItem;
+import ch.epfl.scala.bsp4j.OutputPathsItem;
+import ch.epfl.scala.bsp4j.OutputPathsParams;
+import ch.epfl.scala.bsp4j.OutputPathsResult;
+import ch.epfl.scala.bsp4j.SourceItem;
+import ch.epfl.scala.bsp4j.SourcesItem;
+import ch.epfl.scala.bsp4j.SourcesParams;
+import ch.epfl.scala.bsp4j.SourcesResult;
+import ch.epfl.scala.bsp4j.WorkspaceBuildTargetsResult;
+
+public class GradleBuildServerBuildSupport implements IBuildSupport {
+
+    private static final Pattern GRADLE_FILE_EXT = Pattern.compile("^.*\\.gradle(\\.kts)?$");
+    private static final String GRADLE_PROPERTIES = "gradle.properties";
+
+    private static final String OUTPUT_KIND_SOURCE = "source";
+
+    private static final IClasspathAttribute testAttribute = JavaCore.newClasspathAttribute(IClasspathAttribute.TEST, "true");
+    private static final IClasspathAttribute optionalAttribute = JavaCore.newClasspathAttribute(IClasspathAttribute.OPTIONAL, "true");
+
+    @Override
+    public boolean applies(IProject project) {
+        return Utils.isGradleBuildServerProject(project);
+    }
+
+    public void updateClasspath(IProject project, IProgressMonitor monitor) throws CoreException {
+        IPath rootPath = ProjectUtils.findBelongedWorkspaceRoot(project.getLocation());
+        if (rootPath == null) {
+            JavaLanguageServerPlugin.logError("Cannot find workspace root for project: " + project.getName());
+            return;
+        }
+        BuildServer buildServer = ImporterPlugin.getBuildServerConnection(rootPath);
+        List<IClasspathEntry> classpath = new LinkedList<>();
+        WorkspaceBuildTargetsResult workspaceBuildTargetsResult = buildServer.workspaceBuildTargets().join();
+        List<BuildTarget> allBuildTargets = workspaceBuildTargetsResult.getTargets();
+        List<BuildTarget> buildTargets = Utils.getBuildTargetsByProjectUri(allBuildTargets, project.getLocationURI());
+        // put test targets to the end of the list
+        moveTestTargetsToEnd(buildTargets);
+
+        for (BuildTarget buildTarget : buildTargets) {
+            boolean isTest = buildTarget.getTags().contains(BuildTargetTag.TEST);
+            OutputPathsResult outputResult = buildServer.buildTargetOutputPaths(
+                    new OutputPathsParams(Arrays.asList(buildTarget.getId()))).join();
+            String sourceOutputUri = getOutputUriByKind(outputResult.getItems(), OUTPUT_KIND_SOURCE);
```

Comment:
Why do we need to check if it's a source output? Is there any case where the output is not from source?

Reply:
Output for source and resource are different in gradle: https://github.com/microsoft/build-server-for-gradle/blob/eb2f1b7c51ded39d440a1ce82c2c41e9d397f96a/server/src/main/java/com/microsoft/java/bs/core/internal/services/BuildTargetService.java#L154-L168

Unfortunately, in BSP, it does not distinguish different kind of output. I'm asking them the best way to handle this: https://github.com/build-server-protocol/build-server-protocol/issues/582. No answer so far.